### PR TITLE
[Merged by Bors] - feat: Sublattices

### PR DESCRIPTION
**Pull request by YaelDillies**
Basic API for sublattices (there's more to come).

---

- [ ] depends on: #7548

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by jcommelin — APPROVED** [review]
Thanks :tada: 

If CI passes, please remove the label `awaiting-CI` and merge this yourself, by adding a comment `bors r+`.

bors d+

File: Mathlib/Order/Sublattice.lean [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+variable {L M : Sublattice α} {f : LatticeHom α β} {s t : Set α} {a : α}
+
+initialize_simps_projections Sublattice (carrier → coe)
+
+instance instSetLike : SetLike (Sublattice α) α where
+  coe L := L.carrier
+  coe_injective' L M h := by cases L; congr
+
+/-- Turn a set closed supremum and infimum into a sublattice. -/
+def ofIsSublattice (s : Set α) (hs : IsSublattice s) : Sublattice α := ⟨s, hs.1, hs.2⟩
```
Comment (eric-wieser) [inline Mathlib/Order/Sublattice.lean:46]:
```suggestion
abbrev ofIsSublattice (s : Set α) (hs : IsSublattice s) : Sublattice α := ⟨s, hs.1, hs.2⟩
```
to save you having to write any API for this.

File: Mathlib/Order/Sublattice.lean [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+  { completeLatticeOfInf (Sublattice α)
+      fun _s ↦ IsGLB.of_image SetLike.coe_subset_coe isGLB_biInf with
+    bot := ⊥
+    bot_le := fun _S _a ↦ False.elim
+    top := ⊤
+    le_top := fun _S a _ha ↦ mem_top a
+    inf := (· ⊓ ·)
+    le_inf := fun _L _M _N hM hN _a ha ↦ ⟨hM ha, hN ha⟩
+    inf_le_left := fun _L _M _a ↦ And.left
+    inf_le_right := fun _L _M _a ↦ And.right }
```
Comment (eric-wieser) [inline Mathlib/Order/Sublattice.lean:191]: nit: use `where` notation for les `fun`

Comment (YaelDillies) [reply Mathlib/Order/Sublattice.lean:191]: But `where` doesn't support `with` clauses, does it?

Comment (eric-wieser) [reply Mathlib/Order/Sublattice.lean:191]: It does, using `__ := foo` instead of `foo with`

Comment (YaelDillies) [reply Mathlib/Order/Sublattice.lean:191]: Ah uh that was unexpected syntax.

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**